Feature: 330 careers instagram posts

tbrlpld opened the pull request:
Adds a grid gallery of instagram posts to the careers page type. 

~~The posts consist of an image and a link to the original post on Instagram. This is fully manual though! So images need to be uploaded to Wagtail and the link to the instagram post needs to be added completely manually. This path was chosen because of the complexities in dealing with the Instagram API (e.g. the API would require an access key that needs to be renewed at least every three months).~~ 

Scraping was ruled out because it breaks the Instagram T&C. 

--- 

**Update**

After the comment from @zerolab I switched the approach and implemented the Instagram gallery with the `EmebedBlock`. This makes the editor's job easier as now only the Instagram post URLs need to be added to include the image and a link to the original post in the page.

zerolab commented:
Hey @tbrlpld, 
https://developers.facebook.com/docs/instagram/oembed, https://docs.wagtail.io/en/stable/advanced_topics/embeds.html#facebook-and-instagram (and the [actual finder](nor https://github.com/wagtail/wagtail/blob/stable/2.12.x/wagtail/embeds/finders/instagram.py)) do not mention renewing the access key

Wagtail 2.11+ support the new oEmbed endpoints for Instagram/FB so it would make a lot more sense to make use of them.

chris-lawton commented on tbx/project_styleguide/templates/patterns/atoms/instagram-post/instagram-post.html:
```diff
@@ -0,0 +1,12 @@
+{% load wagtailimages_tags %}
+{% with value as post %}
+<div class="instagram-post">
+    <a href="{{ post.post_url }}">
```
as this anchor tag contains only an image and an SVG it'll need an `aria-label` adding to make it accessible, see https://dequeuniversity.com/rules/axe/4.1/link-name

tbrlpld replied:
I will see if I can extract something useful from the oEmbed version. Otherwise this is going to be something generic like "Instagram post".

chris-lawton commented on tbx/project_styleguide/templates/patterns/atoms/instagram-post/instagram-post.html:
```diff
@@ -0,0 +1,12 @@
+<div class="instagram-post">
+    <a href="{{ post.post_url }}">
+        {% image post.image fill-500x500 as image %}
+        <img class="instagram-post__image" src="{{ image.url }}" loading="lazy" />
```
this could have an `alt` tag added to it

chris-lawton commented on tbx/project_styleguide/templates/patterns/atoms/instagram-post/instagram-post.html:
```diff
@@ -0,0 +1,12 @@
+        {% image post.image fill-500x500 as image %}
+        <img class="instagram-post__image" src="{{ image.url }}" loading="lazy" />
+        <svg class="instagram-post__icon" viewBox="0 0 100 100">
+            <use href="#instagram" />
```
there was a discussion around using `href` over `xlink:href` in slack not too long ago and i think the outcome was that even though it's deprecated, it's still safer to use the latter as there were still some browsers that hadn't included support for it yet, worth double checking though. https://torchbox.slack.com/archives/C032PPZDF/p1596099102050900

chris-lawton commented on tbx/static_src/sass/components/_instagram-gallery.scss:
```diff
@@ -0,0 +1,33 @@
+    width: 100%;
+}
+
+.instagram-post__image {
```
can you these elements inside the initial block declaration e.g. 
```
.instagram-post {
    &__image {
    }
    &__icon {
    }
}
```
less to change if the class is updated in the future and less to type for anyone adding to it  : )

tbrlpld commented:
> Hey @tbrlpld,
> 
> Kudos for the alternative approach.
> 
> https://developers.facebook.com/docs/instagram/oembed, https://docs.wagtail.io/en/stable/advanced_topics/embeds.html#facebook-and-instagram (and the [actual finder](nor https://github.com/wagtail/wagtail/blob/stable/2.12.x/wagtail/embeds/finders/instagram.py)) do not mention renewing the access key
> 
> Wagtail 2.11+ support the new oEmbed endpoints for Instagram/FB so it would make a lot more sense to make use of them. Am I missing something?

No, that sounds good. I guess that is the key difference there to the full API access that once the app is setup and live that the key remains. That is what I had expected from the beginning. 

Sorry, I must have missed that when I was looking at the docs.

tbrlpld commented:
@zerolab Do you know if there already is a Facebook app for the Torchbox website?

zerolab commented:
> @zerolab Do you know if there already is a Facebook app for the Torchbox website?

That, I'm afraid I do not know. You would need to talk to the seo/analytics teams.

tbrlpld commented:
@zerolab I have now implemented the posts with the Wagtail embed block. Thanks for the pointer 👍 Would you mind taking a look at my code? 

@chris-lawton While changing to the Embed version I had to make a few changes to the Frontend code to. I have also tried to incorporate your previous comments. Would you mind taking another look at this?

zerolab reviewed (changes requested):
@tbrlpld Looks/feels much better than previous implementation!
Added a few questions/suggestions, most are not blocking. The main thing I would change is create a custom `InstagramEmbedBlock` derived from `embed_blocks.EmbedBlock` with a url validator and some useful help text

tbrlpld commented on docker-compose.yml:
```diff
@@ -22,7 +22,6 @@ services:
     ports:
       - 8000:8000 # runserver
       - 8001:8001 # mkdocs serve
-      - 6900:6900 # pudb
```
Not a specific decision of mine. I just pulled in the `fabfile.py` and `docker-compose.yml` from `master`. 

https://github.com/torchbox/wagtail-torchbox/commit/189ecbb6a0dc3643847fa3e7b268d4e99c1ab5eb

https://github.com/torchbox/wagtail-torchbox/blame/master/docker-compose.yml

tbrlpld commented on docker-compose.yml:
```diff
@@ -72,8 +71,11 @@ services:
-      - ./media:/app/media:delegated
+      - ./media:/app/media:delegated,rw
+      - ./database_dumps:/app/database_dumps:delegated,rw
+    command: tail -f /dev/null
```
As above, changes are coming from master, but I think without the "do nothing forever", the container would not keep running. The question as to why that is necessary on the utils container should maybe be raised on master. I believe you are right, that is probably not necessary.

tbrlpld commented on tbx/people/templatetags/people_tags.py:
```diff
@@ -0,0 +1,10 @@
+
+
+@register.inclusion_tag("patterns/atoms/instagram-post/instagram-post.html")
+def instagrampost(embed):
```
Fair. I think I was influenced by the navigation tags (e.g. `footernav`).

tbrlpld replied:
The custom template tag is now removed in favour of the custom embed block.

tbrlpld commented:
Thank @zerolab for review! I will be implementing the custom block.

tbrlpld commented on tbx/project_styleguide/templates/patterns/molecules/instagram-gallery/instagram-gallery.html:
```diff
@@ -0,0 +1,6 @@
+{% load wagtailcore_tags people_tags %}
+<div class="instagram-gallery">
+    {% for post in value %}
+        {% instagrampost post %}
```
The custom template tag is now removed in favour of the custom embed block.

tbrlpld commented on tbx/people/blocks.py:
```diff
@@ -33,3 +34,16 @@ def get_link(value):
+class InstagramPostGalleryBlock(blocks.StreamBlock):
+    posts = blocks.StreamBlock(
+        required=False,
+        local_blocks=[("post", embed_blocks.EmbedBlock())],
```
I have implemented the custom block now.

tbrlpld commented:
@zerolab I have implemented the custom embed block. This also removed the need for the template tag, as the block is not defining the rendering. 

Would be great to get you input on this version if you have a moment.

tbrlpld commented:
Thanks for reviewing @chris-lawton 

> I was also getting the following when testing locally, am i putting in an incorrect url?
> 
> ![image](https://user-images.githubusercontent.com/31627284/117802126-f9044f80-b24c-11eb-99ad-e956a5a36846.png)

I suspect the error coming from the fact that you need to configure the embed finder in the settings with an app id and secret. I'll slack you my `local.py` settings.

zerolab commented on tbx/people/blocks.py:
```diff
@@ -33,3 +37,53 @@ def get_link(value):
+        local_blocks=[("post", InstagramEmbedBlock())],
+        min_num=8,
+        max_num=8,
+        template="patterns/molecules/instagram-gallery/instagram-gallery.html",
```
```suggestion
        template="patterns/molecules/instagram-gallery/instagram-gallery.html",
        icon="fa-instagram"
```

zerolab reviewed (approved):
@tbrlpld 👍🏼 

Other than the `.env` question and the `try/except` suggest, the rest are non-blocking suggestions

tbrlpld commented:
@chris-lawton One issue with the focus solution is that I can not rely on the instagram images to be square. When creating the embed, I have no control over what is coming in, but we want them to display as square. That is why I had to mess around with the absolute positioning, etc.

tbrlpld commented:
@chris-lawton can we use `:focus-within` or do we need to support IE11? 

https://caniuse.com/?search=focus-within

tbrlpld commented on .env:
```diff
@@ -0,0 +1 @@
+FRONTEND=local
```
Yup, that should not be committed.

Sorry about that.

tbrlpld commented:
@chris-lawton I was able to solve this more simply by just removing the wrapping `div` and using the `a` element directly as the `instagram-post`.

![Screen Recording 2021-05-21 at 13 03 58](https://user-images.githubusercontent.com/24797493/119192928-b8b78500-ba35-11eb-86ed-928462db9cdc.gif)

tbrlpld commented on tbx/people/blocks.py:
```diff
@@ -33,3 +37,53 @@ def get_link(value):
+    """Custom Embed value that allow access to represented embed object."""
+
+    def embed(self):
+        return embeds.get_embed(self.url)
```
Thanks for the hint @zerolab! 

I have implemented this in 72f29ed6e95407ced1b6fb26d7606b4c31bde103

chris-lawton commented:
Nice work @tbrlpld! LGTM

zerolab reviewed (approved):
Nice one @tbrlpld!

tbrlpld commented:
Thanks @chris-lawton and @zerolab for reviewing
